rasa/utils/tensorflow/models.py

```diff
@@ -451,7 +455,9 @@ def batch_to_model_data_format(
         if isinstance(batch[0], Tuple):
             batch = batch[0]
 
-        batch_data = defaultdict(lambda: defaultdict(list))
+        batch_data: Dict[Text, Dict[Text, List[tf.Tensor]]] = defaultdict(
```

Comment:
We should maybe also open an issue to get rid of default dict here. The `defaultdict`s could have funny side effects. At least I would not expect a dafultdict here. ... Actually confuses me why this doesn't take care of what we solved with `setdefault` for the issue 🤔 .

Reply:
Update: it does, but it is not present everywhere and hence I consistently added "setdefault" everywhere instead of relying on this setdefault

Reply:
>. it does, but it is not present everywhere

Can you elaborate on this?

I am still wondering about this because everywhere `tf_batch_data` and `self.tf_label_data` are constructed from `batch_to_model_data_format` and hence if the defaults are set correctly in this function then it shouldn't be different from setting the defaults to `[]` in other functions. What am I missing? 😅

Reply:
Oh, good point. Will need to disable the setdefaults and check where that fails then. That is indeed strange.

Reply:
done - workaround is: removal check in get_sequence_length functions, renaming of that function to justify the removal of checks, and addition of comments about defaultdicts